parcels/spatialhash.py

```diff
@@ -0,0 +1,326 @@
+
+    def _hash_index2d(self, coords):
+        """Computes the 2-d hash index (i,j) for the location (x,y), where x and y are given in spherical
+        coordinates (in degrees)
+        """
+        # Wrap longitude to [-180, 180]
+        lon = (coords[:, 0] + 180.0) % (360.0) - 180.0
+        i = ((lon - self._xmin) / self._dh).astype(np.int32)
+        j = ((coords[:, 1] - self._ymin) / self._dh).astype(np.int32)
+        return i, j
+
+    def _hash_index(self, coords):
+        """Computes the flattened hash index for the location (x,y), where x and y are given in spherical
+        coordinates (in degrees). The single dimensioned hash index orders the flat index with all of the
+        i-points first and then all the j-points.
+        """
+        i, j = self._hash_index2d(coords)
+        return i + self._nx * j
+
+    def _grid_ij_for_eid(self, eid):
+        """Returns the (i,j) grid coordinates for the given element id (eid)"""
+        j = eid // (self._source_grid.xdim)
+        i = eid - j * (self._source_grid.xdim)
+        return i, j
+
+    def _initialize_face_hash_table(self):
+        """Create a mapping that relates unstructured grid faces to hash indices by determining
+        which faces overlap with which hash cells
+        """
+        if self._face_hash_table is None or self.reconstruct:
+            index_to_face = [[] for i in range(self._nx * self._ny)]
+            # Get the bounds of each curvilinear faces
+            lon_bounds, lat_bounds = curvilinear_grid_facebounds(
+                self._source_grid.lon,
+                self._source_grid.lat,
+            )
+            coords = np.stack(
+                (
+                    lon_bounds[:, :, 0].flatten(),
+                    lat_bounds[:, :, 0].flatten(),
+                ),
+                axis=-1,
+            )
+            xi1, yi1 = self._hash_index2d(coords)
+            coords = np.stack(
+                (
+                    lon_bounds[:, :, 1].flatten(),
+                    lat_bounds[:, :, 1].flatten(),
+                ),
+                axis=-1,
+            )
+            xi2, yi2 = self._hash_index2d(coords)
+            nface = (self._source_grid.xdim) * (self._source_grid.ydim)
+            for eid in range(nface):
+                for j in range(yi1[eid], yi2[eid] + 1):
+                    if xi1[eid] <= xi2[eid]:
+                        # Normal case, no wrap
+                        for i in range(xi1[eid], xi2[eid] + 1):
+                            index_to_face[(i % self._nx) + self._nx * j].append(eid)
+                    else:
+                        # Wrap-around case
+                        for i in range(xi1[eid], self._nx):
+                            index_to_face[(i % self._nx) + self._nx * j].append(eid)
+                        for i in range(0, xi2[eid] + 1):
+                            index_to_face[(i % self._nx) + self._nx * j].append(eid)
+            return index_to_face
+
+    def query(
```

Comment:
Would it be benefial some of the stuff (xbound etc.) that is just dependent on the grid shape so that this isn't computed each query?

Reply:
Since this is only related to spatial hashing, I think that this would be best to be stored on the spatialhash object itself (in whatever format you would like).

Reply:
Done. Resolved in https://github.com/OceanParcels/Parcels/pull/2132/commits/3e2d18eb7bf16e8fe7d708a03f0b50bb0d827d8d